### PR TITLE
fix: Adds .env file detection in compose

### PR DESCRIPTION
**Pull request by seshapad**
Signed-off-by: Padmanabha V Seshadri <seshapad@in.ibm.com>

### CODE REVIEW

**Review by ashokponkumar — CHANGES_REQUESTED** [review]

File: transformer/compose/v3.go [hunk 1]
```diff
@@ -80,6 +81,21 @@ func removeNonExistentEnvFilesV3(path string, parsedComposeFile map[string]inter
...
 						}
+					} else {
+						checkForDotEnv = true
+					}
+				} else {
+					checkForDotEnv = true
+				}
+				if checkForDotEnv {
+					vals := val.(map[string]interface{})
+					envFilePath := filepath.Join(composeFileDir, ".env")
```
[the remark this answers is not in the document]
Comment (seshapad) [reply transformer/compose/v3.go]: done

File: transformer/compose/v1v2.go [hunk 2]
```diff
@@ -95,9 +95,11 @@ func parseV2(path string, interpolate bool) (*project.Project, error) {
 	//TODO: Check if any variable is mandatory
 	someEnvFilePath := ".env"
 	if !common.IgnoreEnvironment {
-		absSomeEnvFilePath, err := filepath.Abs(someEnvFilePath)
+		composeFileDir := filepath.Dir(path)
+		absSomeEnvFilePath := filepath.Join(composeFileDir, someEnvFilePath)
+		_, err := os.Stat(absSomeEnvFilePath)
 		if err != nil {
-			logrus.Errorf("Failed to make the path %s absolute. Error: %q", someEnvFilePath, err)
+			logrus.Errorf("Failed to find the env path %s. Error: %q", someEnvFilePath, err)
```
[the remark this answers is not in the document]
Comment (seshapad) [reply transformer/compose/v1v2.go]: done

File: transformer/compose/v1v2.go [hunk 3]
```diff
@@ -93,11 +93,13 @@ func parseV2(path string, interpolate bool) (*project.Project, error) {
...
+	someEnvFilePath := defaultEnvFile
 	if !common.IgnoreEnvironment {
-		absSomeEnvFilePath, err := filepath.Abs(someEnvFilePath)
+		composeFileDir := filepath.Dir(path)
+		absSomeEnvFilePath := filepath.Join(composeFileDir, someEnvFilePath)
+		_, err := os.Stat(absSomeEnvFilePath)
 		if err != nil {
-			logrus.Errorf("Failed to make the path %s absolute. Error: %q", someEnvFilePath, err)
+			logrus.Debugf("Failed to find the env path %s. Error: %q", someEnvFilePath, err)
 			return nil, err
```
Comment (ashokponkumar) [inline transformer/compose/v1v2.go:103]: Why do we want to return with an error when the .env file is not found?

File: transformer/compose/v3.go [hunk 4]
```diff
@@ -80,6 +81,21 @@ func removeNonExistentEnvFilesV3(path string, parsedComposeFile map[string]inter
 							}
 							vals[envFile] = envfiles
 						}
+					} else {
+						checkForDotEnv = true
+					}
+				} else {
+					checkForDotEnv = true
+				}
+				if checkForDotEnv {
```
Comment (ashokponkumar) [inline transformer/compose/v3.go:90]: Does Docker compose too use the `.env` file only if there is no environment file specified?

File: transformer/compose/v3.go [hunk 5]
```diff
@@ -80,6 +81,21 @@ func removeNonExistentEnvFilesV3(path string, parsedComposeFile map[string]inter
...
+					vals := val.(map[string]interface{})
+					envFilePath := filepath.Join(composeFileDir, defaultEnvFile)
+					finfo, err := os.Stat(envFilePath)
+					if os.IsNotExist(err) || finfo.IsDir() {
+						logrus.Debugf("Unable to find env config file %s referred in service %s in file %s. Ignoring it.", envFilePath, serviceName, path)
```
Comment (ashokponkumar) [inline transformer/compose/v3.go:95]: If we are here, it means it is not referenced in the docker compose file, isn't it?

File: transformer/compose/v3.go [hunk 6]
```diff
@@ -80,6 +81,21 @@ func removeNonExistentEnvFilesV3(path string, parsedComposeFile map[string]inter
...
+					} else {
+						vals[envFile] = defaultEnvFile
+						logrus.Debugf("env config file %s referred in service %s in file %s. Adding  it.", envFilePath, serviceName, path)
```
Comment (ashokponkumar) [inline transformer/compose/v3.go:98]: If we are here, it means it is not referenced in the docker compose file, isn't it?

**Review by ashokponkumar — CHANGES_REQUESTED** [review]

File: transformer/compose/v3.go (hunk 5 above)
Comment (seshapad) [reply transformer/compose/v3.go:95]: if we hit this condition, it means that docker compose has not listed any env file. So, as a follow-on, we also check if .env is available or not. If available we add it to the env list, if the file does not exist, then we dont add anything.

File: transformer/compose/v3.go (hunk 6 above)
Comment (seshapad) [reply transformer/compose/v3.go:98]: if we hit this condition, it means that docker compose has not listed any env file. So, as a follow-on, we also check if .env is available or not. If available we add it to the env list, if the file does not exist, then we dont add anything.

File: transformer/compose/v1v2.go (hunk 3 above)
Comment (seshapad) [reply transformer/compose/v1v2.go]: Removed this return.

File: transformer/compose/v3.go (hunk 5 above)
Comment (ashokponkumar) [inline transformer/compose/v3.go:95]:
```suggestion
						logrus.Debugf("Unable to find env file %s for service %s in file %s. Ignoring it.", envFilePath, serviceName, path)
```

File: transformer/compose/v3.go (hunk 6 above)
Comment (ashokponkumar) [inline transformer/compose/v3.go:98]:
```suggestion
						logrus.Debugf("env file %s for service %s in file %s. Adding  it.", envFilePath, serviceName, path)
```

**Review by ashokponkumar — CHANGES_REQUESTED** [review]

File: transformer/compose/v1v2.go [hunk 7]
```diff
@@ -93,12 +108,13 @@ func parseV2(path string, interpolate bool) (*project.Project, error) {
...
-		absSomeEnvFilePath, err := filepath.Abs(someEnvFilePath)
+		composeFileDir := filepath.Dir(path)
+		absSomeEnvFilePath := filepath.Join(composeFileDir, someEnvFilePath)
+		_, err := os.Stat(absSomeEnvFilePath)
 		if err != nil {
-			logrus.Errorf("Failed to make the path %s absolute. Error: %q", someEnvFilePath, err)
-			return nil, err
+			logrus.Debugf("Failed to find the env path %s. Error: %q", someEnvFilePath, err)
 		}
 		someEnvFilePath = absSomeEnvFilePath
```
[the remark this answers is not in the document]
Comment (seshapad) [reply transformer/compose/v1v2.go]: Done.

**Review by ashokponkumar — APPROVED** [review]

**Pull request merged**